R/tmb.R

```diff
@@ -1,183 +1,39 @@
-#' - `full_formula`: same as `model_formula` but includes the `subject_var` and
-#'      `visit_var`.
+#' - `model_formula`: `formula` with the covariance term removed.
+#' - `full_formula`: same as `model_formula` but includes the covariance
+#'   structure's subject, visit and (optionally) group variables.
 #' - `cov_type`: `string` with covariance term type (e.g. `"us"`).
-#' - `is_spatial`: `flag` indicator of whether the covariance structure is spatial
+#' - `is_spatial`: `flag` indicator of whether the covariance structure is
+#'   spatial
 #' - `visit_var`: `character` with the visit variable name.
 #' - `subject_var`: `string` with the subject variable name.
-#' - `group_var`: `string` with the group variable name. If no group specified, this element
-#'      is `NULL`.
+#' - `group_var`: `string` with the group variable name. If no group specified,
+#'   this element is `NULL`.
+#'
 #' @keywords internal
-h_mmrm_tmb_formula_parts <- function(formula) {
+h_mmrm_tmb_formula_parts <- function(formula, covariance) {
   assert_formula(formula)
-  # Ensure that there is left and right hand side in the formula.
   assert_true(identical(length(formula), 3L))
 
-  # Find the covariance specification term in the formula.
-  cov_functions <- c(cov_type, cov_type_spatial)
-  terms_object <- stats::terms(formula, specials = cov_functions)
-  found_specials <- attr(terms_object, "specials")
-  cov_selected <- Filter(Negate(is.null), found_specials)
-  if (length(cov_selected) == 0) {
-    stop(
-      "Covariance structure must be specified in formula. ",
-      "Possible covariance structures include: ",
-      toString(cov_functions)
-    )
-  } else if (length(cov_selected) > 1) {
-    stop(
-      "Only one covariance structure can be specified. ",
-      "Currently specified covariance structures are: ",
-      toString(names(cov_selected))
-    )
-  }
-  terms_list <- attr(terms_object, "variables")
-  cov_term <- terms_list[[unlist(cov_selected) + 1L]]
-  # Remove the covariance term to obtain the model formula.
-  model_formula <- stats::update(
-    formula,
-    stats::as.formula(paste(". ~ . -", deparse(cov_term)))
-  )
-  cov_vars <- h_mmrm_tmb_extract_vars(cov_term)
-  full_formula <- stats::update(
-    model_formula,
-    stats::as.formula(paste(". ~ . +", cov_vars$subject_var, "+", paste0(cov_vars$visit_var, collapse = "+")))
-  )
-  if (!is.null(cov_vars$group_var)) {
-    full_formula <- stats::update(
-      full_formula,
-      stats::as.formula(paste(". ~ . +", cov_vars$group_var))
-    )
-  }
-  cov_term_type <- deparse(cov_term[[1L]])
-  if (length(cov_vars$visit_var) > 1L && !cov_term_type %in% cov_type_spatial) {
-    stop(
-      "Only spatial covariance support multiple coordinates"
-    )
-  }
+  full_formula <- add_covariance_variable_terms(formula, covariance)
+
   structure(
     list(
-      formula = formula,
-      model_formula = model_formula,
+      # formula = formula,
+      model_formula = formula,
```

Comment:
I guess we need to be careful here that we don't break anything downstream

Reply:
I scanned the repo and the first `formula` was never used. Still need to confirm in testing, but I'd prefer to prune some dead code if we can.